src/environment.cpp

```diff
@@ -1544,6 +1544,49 @@ void ServerEnvironment::step(float dtime)
 		*/
 		removeRemovedObjects();
 	}
+
+	/*
+		Manage particle spawner expiration
+	*/
+	if (m_particle_management_interval.step(dtime, 1.0)) {
+		for (std::map<u32, float>::iterator i = m_particle_spawners.begin();
+				i != m_particle_spawners.end(); ) {
+			//non expiring spawners
+			if (i->second <= -1.f) {
+				++i;
+				continue;
+			}
+
+			i->second -= 1.0f;
+			if (i->second <= 0.f)
+				m_particle_spawners.erase(i++);
+			else
+				++i;
+		}
+	}
+}
+
+u32 ServerEnvironment::addParticleSpawner(float exptime)
+{
+	// Timers with lifetime 0 do not expire
+	float time = exptime > 0.f ? exptime : -1.f;
```

Comment:
If this is supposed to be non-expiring it needs to be something other than -1.0f (-2.0 will do, and better make it a constant or define) and line 1555 changed to if (i->second < -1.0f)

If this doesn't make sense consider an expiring particle spawner starting with a lifetime of 2.0f. It's "lifetime" value sequence will be 2.0f, 1.0f, 0.0f, -1.0f. At that point it becomes indistinguishable from a non-expiring particle spawner and never be deleted (because of the condition currently on line 1555).

Summary: This line needs to change the initial time to something less than -1.0f and line 1555 needs to be updated to delete spawners with a lifetime of -1.0f (i.e. change to < not <=)

NB This has nothing to do with floating point accuracy or precision and for the purpose of this comment is irrelevant